maro/simulator/scenarios/supply_chain/sku_dynamics_sampler.py

```diff
@@ -0,0 +1,294 @@
+        self._fp = open(self._preprocessed_file_path, "rt")
+        self._reader = DictReader(self._fp)
+        self._is_fp_closed = False
+        self._latest_tick = None
+
+    def _load_data_until_tick(self, tick: int) -> None:
+        """Load all data that are no later than `tick`
+
+        May load one more extra entry to ensure that all data at tick `tick` are loaded.
+        """
+        while not self._is_fp_closed and (self._latest_tick is None or self._latest_tick <= tick):
+            row = next(self._reader)  # This entry may be after `tick`. We have to process it as we already loaded it.
+            if row is None:
+                self._fp.close()
+                self._is_fp_closed = True
+                break
+
+            date_index = int(row[DATE_INDEX_COLUMN_NAME])
+
+            if self._start_date_index is None:
+                self._start_date_index = date_index
+
+            # So one day one tick.
+            self._latest_tick = target_tick = date_index - self._start_date_index
+
+            sku_name = row[self._sku_column_name]
+            if sku_name not in self._world.sku_name2id_mapping:
+                continue
+            sku_id = self._world.sku_name2id_mapping[sku_name]
+
+            self._cache[target_tick][sku_id] = {}
+            for attr_name, (column_name, type_name) in self._info_dict.items():
+                self._cache[target_tick][sku_id][attr_name] = type_name(row[column_name])
+
+
+class SkuPriceInterface(metaclass=ABCMeta):
+    """Price sample interface."""
+
+    @abstractmethod
+    def sample_price(self, tick: int, product_id: int) -> Optional[float]:
+        """Sample the price for specified product and tick.
+
+        Args:
+            tick (int): Tick of environment, NOTE: this tick is start from 0,
+                you may need to transform it to your time system.
+            product_id (int): Id of product to sample.
+
+        Return:
+            Optional[float]: the new price in the specific tick. None if not changed.
+        """
+        raise NotImplementedError
+
+
+class OneTimeSkuPriceDemandSampler(OneTimeSkuDynamicsSampler, SkuPriceInterface, SellerDemandInterface):
+    def _init_info_dict(self) -> Dict[str, DynamicsInfoItem]:
+        return {
+            "Price": DynamicsInfoItem(self._configs.get("price_column", "Price"), float, None),
+            "Demand": DynamicsInfoItem(self._configs.get("demand_column", "Demand"), int, 0),
+        }
+
+    def _sample_attr(self, tick: int, product_id: int, attr_name: str) -> object:
+        if any([
+            tick not in self._cache,
+            product_id not in self._cache[tick],
+            attr_name not in self._cache[tick][product_id],
+        ]):
+            return self._info_dict[attr_name].default_value
+
+        return self._cache[tick][product_id][attr_name]
+
+    def sample_price(self, tick: int, product_id: int) -> Optional[float]:
```

Comment:
Currently the IDE is confused about the return type, since the lower level method returns an `object`.

To my personal habit, I will add an type check to explicitly tell the IDE the type of the returned value, such as:

```python
    def sample_price(self, tick: int, product_id: int) -> Optional[float]:
        ret = self._sample_attr(tick, product_id, "Price")
        assert isinstance(ret, float)
        return ret
```

It is optional. Same for `sample_demand`